Add alternate enable config for spindle control

benagricola opened the pull request:
Discussed here - https://forum.duet3d.com/topic/35936/direction-control-in-rrf/6?_=1730375174481 - I have been using this implementation on my own system for a couple of months and it works perfectly.

For directional spindle control, RRF currently has 2 pins - `enable` and `direction`, with the following states:

```
SpindleState::forward: enable=1, direction=0
SpindleState::reverse: enable=1, direction=1
SpindleState::stopped: enable=0, direction=? // Unknown, as RRF does not change direction pin state when stopping spindle
```

This requires connecting to a VFD that expects an `enable` pin and a `direction` pin, where it is valid for both enable and direction to be logic high at once (and that means reverse).

There are lots of VFDs which do not use this control mechanism, instead taking a `forward` and `reverse` input. If you connect RRF to one of these VFDs, (`enable` => `forward` and `direction` => `reverse`) you will likely only have forward-run capability - trying to switch to reverse run will stop the spindle because both forward and reverse inputs are high.

_However_ - if the spindle is in reverse in RRF and you then _stop_ it, there is a possibility that the VFD will start the spindle at that point - because RRF does not set the state of the direction pin when stopping the spindle.

This is obviously not ideal and means RRF cannot be used safely for reverse running with VFDs that don't match the RRF Spindle control mechanism (which isn't necessarily obvious).

I propose in this PR an alternate configuration, using the alternate pin specifier `*` on the `enable` pin. This would be given as part of the `M950` command to configure the spindle, for example:

```gcode
; Configure spindle in Normal mode
; Forward: spindleen=1, spindledir=0
; Backward: spindleen=1, spindledir=1
M950 R0 C"spindlepwm+spindleen+spindledir" L1999:24000 Q1000

; Configure spindle in Alternate mode
; Forward: spindleen=1, spindledir=0
; Backward: spindleen=0, spindledir=1
M950 R0 C"spindlepwm+*spindleen+spindledir" L1999:24000 Q1000
```

This PR implements this alternate behaviour, and also makes sure that the `direction` pin has its' value set when the spindle is stopped to avoid accidentally starting the spindle when it should be stopped.

dc42 commented:
Thanks for your contribution. However, I am not happy with adding a * character to the pin name to change the behaviour. I would rather use a new parameter to specify the controller type. I suggest using parameter T (for Type) which is already used by some other variants of M950. We can use T0 to specify the existing behaviour, and T1 to specify the new behaviour that you desire. This leaves T2, 3.. available for future additional control types, in particular we intend to support Modbus RTU control of spindles in the future.

Please revise your PR to use this new parameter. If the T parameter is omitted then it should default to 0 to behave as now. The value of the T parameter should be restricted to 0 or 1, which you can do easily by using function GCodeBuffer::TryGetLimitedUIValue to test for the parameter and read its value. M950 Rn with no additional parameters should include the type of the spindle in the returned information. The type of the spindle should be added to its description in the object model.

benagricola committed:
Add spindle type, std and fwdrev control schemes

Spindles now have a type, which defaults to standard (`T0`).
In standard mode, spindles have an enable and a direction port:

  Forward: enable=1, direction=0
  Reverse: enable=1, direction=1

Setting fwdrev (`T1`)  mode on the spindle changes the ports
to represent forward and reverse:

  Forward: forward=1, reverse=0
  Reverse: forward=0, reverse=1

Signed-off-by: Ben Agricola <ben+git@agrico.la>

benagricola commented:
Hi @dc42 - I've reimplemented this given your feedback. This PR now accepts a `T0` or `T1` parameter to `M950 R...`, allowing the spindle type to be configured.

The type field is used to change the behaviour of the `onOffPort` when the spindle is in reverse.

I've implemented display of the configured spindle when called without any parameters, printing the spindle type as well as pin and rpm details.

Finally I've implemented the `type` field in the object model - I couldn't find a way to add this to the API spec - I assume this is done separately?

Unsure of the `type` names but `standard` and `fwdrev` are what came to mind so if there's any better suggestions here I'm all ears.

![spindletype-0](https://github.com/user-attachments/assets/eb6c9dcc-f2f8-4996-b5f6-b3c935152610)
![spindletype-1](https://github.com/user-attachments/assets/2a2a99bb-080e-4095-86d7-33de809318bc)

benagricola committed:
Adjust return logic to reduce nesting

Signed-off-by: Ben Agricola <ben+git@agrico.la>
